quinn-proto/src/connection/paths.rs

```diff
@@ -87,3 +91,99 @@ impl RttEstimator {
             .map(|srtt| srtt + cmp::max(4 * self.var, TIMER_GRANULARITY))
     }
 }
+
+// Implements Datagram Packetization Layer Path Maximum Transmission Unit Discovery
+//
+// https://www.ietf.org/id/draft-ietf-tsvwg-datagram-plpmtud-21.html
+pub struct MtuDiscovery {
+    header_size: u16,
+    // Current MTU for the path
+    pub current: u16,
+    // Packet number and probe size for the current probe
+    probe_number: Option<u64>,
+    probe_size: Option<u16>,
+    // Failed probes at the current probe size
+    probe_count: usize,
+    phase: Phase,
+}
+
+impl MtuDiscovery {
+    fn new(remote: SocketAddr) -> Self {
+        Self {
+            header_size: match remote {
+                SocketAddr::V4(_) => 20,
+                SocketAddr::V6(_) => 48,
+            },
+            current: BASE_PLPMTU,
+            probe_number: None,
+            probe_size: None,
+            probe_count: 0,
+            phase: Phase::Searching,
+        }
+    }
+
+    pub fn poll_transmit(&mut self, next_packet_number: u64) -> Option<u16> {
+        if self.probe_number.is_some() {
+            return None;
+        } else if let Phase::Complete = self.phase {
+            return None;
+        }
+
+        if self.probe_size.is_none() {
+            match LEVELS
+                .iter()
+                .find(|&&x| x > (self.current + self.header_size))
+            {
+                Some(v) => {
+                    self.probe_size = Some(*v);
+                }
+                None => {
+                    self.phase = Phase::Complete;
+                    return None;
+                }
+            }
+        }
+
+        self.probe_number = Some(next_packet_number);
+        self.probe_size
+    }
+
+    pub fn acked(&mut self, number: u64) {
+        match self.probe_number {
+            Some(probed) if probed == number => {}
+            _ => return,
+        };
+
+        self.probe_number = None;
+        let new = self.probe_size.take().unwrap();
+        self.current = new - self.header_size;
+        if self.current == MAX_PLPMTU {
+            self.phase = Phase::Complete;
+        }
+    }
+
+    pub fn lost(&mut self, number: u64) {
+        match self.probe_number {
+            Some(probed) if probed == number => {}
+            _ => return,
+        };
+
+        self.probe_number = None;
+        self.probe_count += 1;
+        if self.probe_count == MAX_PROBES {
+            self.probe_size = None;
+            self.phase = Phase::Complete;
+        }
+    }
+}
+
+enum Phase {
+    Searching,
+    Complete,
+}
+
+const LEVELS: [u16; 4] = [1_350, 1_400, 1_450, 1_500];
```

Comment:
The spec has this bit:

"The search algorithm determines a minimum useful gain in PLPMTU. It would not be constructive for a PL sender to attempt to probe for all sizes. This would incur unnecessary load on the path. Implementations SHOULD select the set of probe packet sizes to maximize the gain in PLPMTU from each search step.

Implementations could optimize the search procedure by selecting step sizes from a table of common PMTU sizes. When selecting the appropriate next size to search, an implementer ought to also consider that there can be common sizes of MPS that applications seek to use, and their could be common sizes of MTU used within the network."

I think the hardcoded search space is nice and straightforward, and there's not much benefit to making it more continuous. Also makes it easy to modify/experiment with the step changes. The `Phase` enum is trivial but also clearly links up to the spec, which I think is helpful and nicely explicit even if a `bool` could otherwise do.

Reply:
I don't think it needs to be continuous, but it'd be nice to support transports with larger MTUs, e.g. LANs with jumbo frames enabled. Given a significantly larger maximum MTU, a constant increment also won't make sense, since we probably want to find the limit faster than a linear search.

For a first implementation, I think this simple approach based on the realities of the internet is great, but I do think there's a place here for a more powerful approach eventually.